Add foundation for E2E Tests + First E2E Scenario: Register Orchestrator

leszko opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Create first E2E Scenario which registers an orchestrator using local chain.

How to run the test:
1. You need to have Docker running
2. (Mac M1) Update `Image` in `e2e.go` to `darkdragon/geth-with-livepeer-protocol:streamflow`
3. Fire `./test_e2e.sh`

For more details about next steps, please check [E2E Tests Epic](https://github.com/livepeer/go-livepeer/issues/2297)

**Specific updates (required)**
- Extract `devtool` package from `devtool.go` to make the devtool code reusable
- Extract `starter` package from `livepeer.go` to make the livepeer code reusable
- Create `e2e.go` with the reusable utils for E2E Tests
- Create `register_orchestrator_test.go` with the first E2E Test scenario
- Create/Update `test.sh`/`test_e2e.sh`

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->
- `./test-e2e.sh`
- check if nothing is broken with devtool
- check if nothing is broken with livepeer


**Does this pull request close any open issues?**
fix #2317


**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [ ] ~~README and other documentation updated~~
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

thomshutt commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,223 @@
+		ContainerRequest: req,
+		Started:          true,
+	})
+	if err != nil {
```
Can trim the code down a bit with `require.NoError(t, err)` rather than this conditional pattern

leszko replied:
Fixed in 60644542ad9d41ae34f04b92a7b126cd08e1da3a

thomshutt commented:
If the devtool_utils are only going to be used in the tests then could we move them over to some more specific files/packages within the E2E code?

thomshutt commented:
Not specific to this PR, but we should look at moving the M1 image to a livepeer-owned repo

thomshutt commented:
I tried running the tests (M1 MacOS. I updated the image name to the darkdragon one) and it seems to hang after this point:

```
++ go list ./...
++ grep test/e2e
+ go test github.com/livepeer/go-livepeer/test/e2e
# github.com/karalabe/usb
In file included from ../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libs.go:50:
../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libusb/libusb/os/darwin_usb.c:253:39: warning: 'kIOMasterPortDefault' is deprecated: first deprecated in macOS 12.0 [-Wdeprecated-declarations]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/IOKit.framework/Headers/IOKitLib.h:123:19: note: 'kIOMasterPortDefault' has been explicitly marked deprecated here
In file included from ../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libs.go:50:
../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libusb/libusb/os/darwin_usb.c:390:26: warning: 'kIOMasterPortDefault' is deprecated: first deprecated in macOS 12.0 [-Wdeprecated-declarations]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/IOKit.framework/Headers/IOKitLib.h:123:19: note: 'kIOMasterPortDefault' has been explicitly marked deprecated here
In file included from ../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libs.go:50:
../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libusb/libusb/os/darwin_usb.c:441:60: warning: 'kIOMasterPortDefault' is deprecated: first deprecated in macOS 12.0 [-Wdeprecated-declarations]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/IOKit.framework/Headers/IOKitLib.h:123:19: note: 'kIOMasterPortDefault' has been explicitly marked deprecated here
In file included from ../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/libs.go:51:
../../../../pkg/mod/github.com/karalabe/usb@v0.0.0-20211005121534-4c5740d64559/hidapi/mac/hid.c:693:34: warning: 'kIOMasterPortDefault' is deprecated: first deprecated in macOS 12.0 [-Wdeprecated-declarations]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/IOKit.framework/Headers/IOKitLib.h:123:19: note: 'kIOMasterPortDefault' has been explicitly marked deprecated here
# github.com/livepeer/lpms/ffmpeg
decoder.c:234:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
```

After a while, it eventually errors:

```
2022/04/29 17:43:49 Starting container id: a9fad7a881cf image: testcontainers/ryuk:0.3.3
2022/04/29 17:43:49 Waiting for container id a9fad7a881cf image: testcontainers/ryuk:0.3.3
2022/04/29 17:43:49 Container is ready id: a9fad7a881cf image: testcontainers/ryuk:0.3.3
2022/04/29 17:43:49 Starting container id: 3787d385e184 image: darkdragon/geth-with-livepeer-protocol:streamflow
2022/04/29 17:43:49 Container is ready id: 3787d385e184 image: darkdragon/geth-with-livepeer-protocol:streamflow
E0429 17:44:12.132515   79915 handlers.go:1392] HTTP Response Error 500: transaction failed txHash=0xbd9c0231620f44cc8712ca367c2317a1c44312eced285885bac996819cfe40c0
E0429 17:44:12.351076   79915 handlers.go:1392] HTTP Response Error 400: orchestrator already registered
```

leszko commented:
> If the devtool_utils are only going to be used in the tests then could we move them over to some more specific files/packages within the E2E code?

It's actually not only used in tests, but also in the [`devtool` cmd](https://github.com/livepeer/go-livepeer/blob/master/cmd/devtool/devtool.go).

leszko commented:
> Not specific to this PR, but we should look at moving the M1 image to a livepeer-owned repo

There is a dedicated GH Issue which will address this one: https://github.com/livepeer/go-livepeer/issues/2316

leszko commented:
> I tried running the tests (M1 MacOS. I updated the image name to the darkdragon one) and it seems to hang after this point:

Right, I had the same. And the reason of that is that I rebased to master and there was a bug in master 😱 
Fixed in 01750faebbc6453a2aa04d3e470d0ca34d0ce5ee.

And for the `master` I'll send a fix in a separate PR in a few hours.

leszko reviewed (commented):
Thanks for the review @thomshutt

I addressed your comments. PTAL

thomshutt commented on test/e2e/register_orchestrator_test.go:
```diff
@@ -0,0 +1,77 @@
+		"amount":         {fmt.Sprintf("%d", lptStake)},
+	}
+
+	for {
```
Will this infinitely loop in the case where there's something wrong with the activation endpoint?

leszko replied:
Good point! Yes, we actually don't use any timeouts in any tests. The full/complete solution we can consider at some point is to use context everywhere with the proper timeout. Saying that it'd slightly complicate the code. I think that for now it's good enough to add `-timeout` to `go test`. Added here: 6acddeb3f6a777be82970cb0e87dd086b7890b28

thomshutt commented on cmd/devtool/devtool/devtool_utils.go:
```diff
@@ -0,0 +1,370 @@
+func (d *Devtool) InitializeRound() error {
+	// XXX TODO curl -X "POST" http://localhost:$transcoderCliPort/initializeRound
+	time.Sleep(7 * time.Second)
+	for {
```
Same here, should we have a timeout?

leszko replied:
Same as above ⬆️

thomshutt commented on cmd/devtool/devtool/devtool_utils.go:
```diff
@@ -0,0 +1,370 @@
+	tx, err := d.Client.InitializeRound()
+	// ErrRoundInitialized
+	if err != nil {
+		if err.Error() != "ErrRoundInitialized" {
```
Do we need to handle the case where `err != nil` but it's not this type of error?

leszko replied:
No, if the error is `"ErrRoundInitialized"`, then no problem. We don't need to initialize it, because it's already initialized (potentially by someone else).

Saying that, in `devtool_util.go`, there are a lot of places which could be improved in terms of the code quality, but I don't want to do it this PR. Maybe we can send some following PR, just with the refactoring.

thomshutt replied:
My bad, I'd read `err.Error() != "ErrRoundInitialized"` as `err.Error() == "ErrRoundInitialized"`

thomshutt commented:
Overall looks great, really nice and neat - I played around with making the test fail and added another (duplicate) test to try and check there weren't any race conditions around setup and teardown etc.

Some general thoughts for the future:

* Since we didn't go with Cucumber as a meta language, we should try to be strict about keeping new tests looking neat and having Given / When / Then comments like `TestRegisterOrchestrator` does
* It might be good to pipe the log outputs of the processes we start into files (either 1 per test or 1 per process) to make reviewing what happened on test failure easier
* The setup / teardown is still a little cumbersome in that you have to remember to wait for the O to be ready and then defer the teardown. We could experiment with grouping tests in the future and doing subtests using the same O with `T.Run` / `T.Parallel` after the setup steps  https://pkg.go.dev/testing#T.Parallel

leszko commented:
Thanks for the review!

> * Since we didn't go with Cucumber as a meta language, we should try to be strict about keeping new tests looking neat and having Given / When / Then comments like `TestRegisterOrchestrator` does
We still plan to consider Cucumber, please check: https://github.com/livepeer/go-livepeer/issues/2331

I think we should convert it to Cucumber (should be quick), send a PR and discuss if we like it more than pure Golang tests.

> * It might be good to pipe the log outputs of the processes we start into files (either 1 per test or 1 per process) to make reviewing what happened on test failure easier'

Yes, we can consider it. TBH I'd start without it, because producing log files also have drawbacks. Maybe let's check in practice if that's really an issue.

> * The setup / teardown is still a little cumbersome in that you have to remember to wait for the O to be ready and then defer the teardown. We could experiment with grouping tests in the future and doing subtests using the same O with `T.Run` / `T.Parallel` after the setup steps  https://pkg.go.dev/testing#T.Parallel

Yes, good idea. Let's wait until we have more E2E Test scenarios and then we can decide if we should run multiple tests (and if yes, then which tests) using the same started Orchestrator.

leszko reviewed (commented):
@thomshutt Addressed your comments. PTAL.

AlexKordic commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,204 @@
+		starter.StartLivepeer(context.TODO(), lpCfg)
+	}()
+
+	ready := make(chan struct{})
+	go func() {
+		statusEndpoint := fmt.Sprintf("http://%s/status", *lpCfg.CliAddr)
+		var statusCode int
+		for statusCode != 200 {
+			time.Sleep(200 * time.Millisecond)
+			resp, err := http.Get(statusEndpoint)
+			if err == nil {
+				statusCode = resp.StatusCode
+			}
+		}
+		ready <- struct{}{}
+	}()
```
Awesome! I was wondering how to get a signal when node is ready.

leszko replied:
Yeah, also for the context, this is probably good enough for E2E Tests, but we plan to add better health checks to Livepeer: 
- #2076
- #1920

AlexKordic reviewed (approved):
Well Done!

leszko committed:
Add backbone for E2E Tests + First E2E Scenario: Register Orchestrator
